### PR TITLE
Implemented validation for templates and handle exceptions

### PR DESCRIPTION
**Pull request by AnuGayan**

### CODE REVIEW

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java [hunk 1]
```diff
@@ -322,26 +321,61 @@ public static void validateTemplatesProperties(RuleTemplate ruleTemplate) throws
...
+                            + ruleTemplate.getUuid());
+                }
 
-    public static void validateBusinessRuleFromTemplate(BusinessRuleFromTemplate businessRuleFromTemplate) throws TemplateManagerException {
-        // todo: implement
-    }
+            }
+        }else {
+            for (Template template : templates){
+                if (!template.getType().isEmpty()){
```
Comment (Anoukh) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: Are we checking if the type is not empty and throwing an exception?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java [hunk 2]
```diff
@@ -322,26 +321,61 @@ public static void validateTemplatesProperties(RuleTemplate ruleTemplate) throws
...
-    public static void validateTemplate(Template template) throws TemplateManagerException {
-        //todo: no need mostly.
-    }
+    public static void validateTemplate(RuleTemplate ruleTemplate) throws TemplateManagerException {
+        Collection<Template> templates = ruleTemplate.getTemplates();
+        /**
+         * Validation for the
+         *
+         * **/
+        if(ruleTemplate.getType().equals(TemplateManagerConstants.TEMPLATE)){
```
Comment (Anoukh) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: Missing spaces? There are several places where formatting is wrong.

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java [hunk 3]
```diff
@@ -158,6 +162,13 @@ public void deleteBusinessRule(String uuid) throws TemplateManagerException { //
         // todo: else: If found Business Rule is from scratch
     }
 
+    public boolean deployBusinessRule(){
+        boolean successfullyDeployed = false;
+
+
```
Comment (mohanvive) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: Why unnecessary line brakes?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java [hunk 4]
```diff
@@ -59,16 +59,18 @@ public TemplateManagerService() {
...
+        boolean isDeployed = false;
         try {
             derivedTemplates = deriveTemplates(businessRuleFromTemplate);
         } catch (TemplateManagerException e) {
             log.error("Error in deriving templates", e);
         }
         String businessRuleUUID = businessRuleFromTemplate.getUuid();
         try {
-            saveBusinessRuleDefinition(businessRuleUUID, businessRuleFromTemplate); // todo : Implement method
+            saveBusinessRuleDefinition(businessRuleUUID, businessRuleFromTemplate,isDeployed); // todo : Implement method
```
Comment (minudika) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: Remove TODOs

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java [hunk 5]
```diff
@@ -322,26 +321,61 @@ public static void validateTemplatesProperties(RuleTemplate ruleTemplate) throws
...
      */
-    public static void validateTemplate(Template template) throws TemplateManagerException {
-        //todo: no need mostly.
-    }
+    public static void validateTemplate(RuleTemplate ruleTemplate) throws TemplateManagerException {
+        Collection<Template> templates = ruleTemplate.getTemplates();
+        /**
+         * Validation for the
+         *
+         * **/
```
Comment (minudika) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: Please fix the comment

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java [hunk 6]
```diff
@@ -322,26 +321,61 @@ public static void validateTemplatesProperties(RuleTemplate ruleTemplate) throws
...
+                    throw new TemplateManagerException("Invalid template. Template type cannot be null in rule " +
+                            "template " +ruleTemplate.getUuid());
+                }
+                if (!template.getType().equals(TemplateManagerConstants.SIDDHI_APP_TEMPLATE_TYPE)|| !template.getType()
+                        .equals(TemplateManagerConstants.GADGET)||!template.getType().equals(TemplateManagerConstants
+                        .DASHBOARD)){
+                    throw new TemplateManagerException("Invalid template. Template type only can be 'siddhiApp'," +
+                            "'gadget' or " +
+                            "'dashboard'" +
+                            " in rule template "+ ruleTemplate.getUuid());
```
Comment (minudika) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: I think we can get this into 2 lines

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java (hunk 4 above)
Comment (ksdperera) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: Code formatting issue 'businessRuleFromTemplate,isDeployed'

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java (hunk 5 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: Fixed by b14dd18f396f799bd32670842869feb2475e778e

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java (hunk 4 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: FIxed by 47e9832228beaa5747e074303ac9d729cc69abdf

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java (hunk 6 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: This was changed by df3bc5a752f0dfcd03393c5f62409814722e720b

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java (hunk 1 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: This was changed by 3674cdddd78974cf7b8aca6a5feb11440c7c499b

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java (hunk 4 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: FIxed by 47e9832

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java [hunk 7]
```diff
@@ -77,6 +79,8 @@ public void createBusinessRuleFromTemplate(BusinessRuleFromTemplate businessRule
                     log.error(e.getMessage(), e);
                 }
             }
+            isDeployed =true;
```
[the remark this answers is not in the document]
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/services/TemplateManagerService.java]: Fixed by 47e9832 and b14dd18f396f799bd32670842869feb2475e778e

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java [hunk 8]
```diff
@@ -322,26 +321,61 @@ public static void validateTemplatesProperties(RuleTemplate ruleTemplate) throws
...
+            }
+        }else {
+            for (Template template : templates){
+                if (!template.getType().isEmpty()){
+                    throw new TemplateManagerException("Invalid template. Template type cannot be empty in rule " +
+                            "template " + ruleTemplate.getUuid());
+                }
+                if (!template.getType().equals(TemplateManagerConstants.SIDDHI_APP_TEMPLATE_TYPE)){
+                    throw new TemplateManagerException("Invalid template. Template type only can be 'siddhiApp' in " +
+                            "rule template "+ruleTemplate.getUuid());
```
[the remark this answers is not in the document]
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/internal/util/TemplateManagerHelper.java]: This was fixed by 3674cdd

**Pull request merged**